### DISCUSSION TITLE
Codeception v5

### DISCUSSION THREAD

**Discussion by Naktibalda** [Codeception/Codeception#6064] [General]
PHPUnit 10 will be released at the beginning of February and there is always a significant demand to support the latest PHPUnit.
On the other hand, if we implement that support in minor version, it may break tests for some of our users, because some functionality they rely on will be removed or changed. So it is a good reason to increment major version and release Codeception 5.0.

Please post what other changes you would like to see implemented on Codeception v5.

Here is what I'm going to do for Codeception v5:

- [x] Merge [the latest branch](https://github.com/Codeception/phpunit-wrapper/tree/10.0) of phpunit-wrapper back to Codeception/Codeception
- [x] Merge "Improved compatibility with PHPUnit 7+" changes made in [codeception/codeception](https://github.com/Codeception/Codeception/pull/5894) and [phpunit-wrapper](https://github.com/Codeception/phpunit-wrapper/pull/77) repos.
- [x] Remove some old PHPUnit compatibility hacks
- [x] Remove support for Symfony 3 (raise the lowest supported version to 4.4)
- [x] Remove support for old versions of php-dotenv 2 and 3 (possibly 4 too).
- [x] Remove support for codeception/stub v2
- [x] Upgrade to the latest major versions of the following dependencies:
  - [x] ~~codeception/specify 0.* to 1.*~~ Removed dependency instead because it wasn't used by tests
  - [x] ~~monolog/monolog 1.* to 2.* (used by ext/Logger.php)~~ Removed dependency instead because Logger has no tests.
  - [x] ~~squizlabs/php_codesniffer 2.* to 3.*~~ Removed because it wasn't part of CI and code style wasn't enforced. It can be readded latter.
- [x] Delete deprecated code
  - [x]  [Codeception\Util\Stub](https://github.com/Codeception/Codeception/blob/4.1/src/Codeception/Util/Stub.php) was deprecated in 2.3.8, 3 years ago
  - [x] [settings: bootstrap:](https://github.com/Codeception/Codeception/commit/55e91e14567c7476bc29e778eda976d336c24f58) setting was deprecated in 3.1.0 in August 2019
- [x] Review all phpunit-wrapper code in src/PHPUnit/.
- [x] Run tests of require-dev modules in Github Actions (this change will be backported to 4.1)
- [ ] Make [codeception/stub](https://github.com/Codeception/stub) compatible with PHPUnit 10.
- [ ] Update files in src/PHPUnit/NonFinal directory with the latest revision of their counterparts in PHPUnit (the only difference was that they have no `final` keyword, because some Codeception classes extend them).
- [ ] Make it work with PHPUnit 10 after final release.
- [ ] Check how much work it is to make with PHPUnit 9. Support it if it is simple (e.g. a few version or method_exists conditions).
- [ ] Repeat this process with PHPUnit 8.

It would be good to support PHP 7.2 in v5, then we can build phar files for PHP 7.2 with Codeception v5 and phar files for PHP 5.6 with Codeception v4.
If we keep building 2 phar files for Codeception v4 and build 1 or 2 new phar files for v5, some creative solution will be required to keep phar self-update functionality working for v4 PHP 7.2 files. Probably the latest version of that file will stay at https://codeception.com/codecept.phar , but Codeception 5 build will be at https://codeception.com/php73/codecept.phar

**Comment by ThomasLandauer** [comment]
> Remove support for Symfony 3 (probably raise the lowest supported version to 4.4)

Symfony 3.4 is supported till November https://symfony.com/releases/3.4, so maybe it's too early to drop it?

**Reply by Naktibalda** [reply]
My intention is to support Codeception 4 for at least one year, so users of Symfony 3 can continue using it.

**Comment by samdark** [comment]
Some references:

- https://phpunit.de/supported-versions.html
- https://www.php.net/supported-versions.php

**Comment by Naktibalda** [comment]
I started working on it: https://github.com/Codeception/Codeception/pull/6067

**Comment by Naktibalda** [comment]
The latest version of TODO list is in my pull request.

**Comment by TavoNiievez** [comment]
maybe it's a good time to:
```php
<?php

declare(strict_types=1); 
```
everywhere.

**Reply by Naktibalda** [reply]
100% is a high bar.

The only place I don't want to change method signatures in this release is interfaces and classes implemented or extended by modules.
It would be good to keep modules compatible with Codeception v4.

**Reply by TavoNiievez** [reply]
@Naktibalda Have you changed your mind about typing the interfaces?

I think that if we type all the interfaces that would be a good excuse to go from PHP 7.1 to PHP 7.3, since what would be intended would be that those new major versions of the modules are used with Codeception 5 that requires PHP 7.3

**Reply by Naktibalda** [reply]
I already agreed that changing interfaces is good idea in https://github.com/Codeception/lib-innerbrowser/pull/39

**Comment by Naktibalda** [comment]
@DavertMik  promissed to change directory structure in 5.0: https://github.com/Codeception/Codeception/issues/5727#issuecomment-716797659

**Comment by DavertMik** [comment]
Ok, i will go with it!

**Comment by Naktibalda** [comment]
Advanced code coverage settings could be merged to 5.0: https://github.com/Codeception/Codeception/pull/5922
@SamMousa didn't want to support PHPUnit 8, so it is a one more reason to set lowest supported of PHPUnit to 9.0
